rancherdesktop/rule.go

```diff
@@ -0,0 +1,101 @@
+package rancherdesktop
+
+import (
+	"fmt"
+	"github.com/Masterminds/semver/v3"
+)
+
+// Rule represents a constraint on which Versions are supported that
+// applies to instances of Rancher Desktop that satisfy specific criteria.
+type Rule struct {
+	Criteria    Criteria
+	Constraints Constraints
+}
+
+// Criteria is the conditions that are used to determine whether a Rule
+// applies for a given client. All parts of Criteria must be satisfied for
+// the Rule to apply to the client.
+type Criteria struct {
+	AppVersion      *semver.Constraints
+	Platform        string
+	Arch            string
+	PlatformVersion *semver.Constraints
+}
+
+// Constraints contains logic that is applied to a Version to determine
+// the value of its Supported key. All parts of Constraints must be satisfied
+// for a Version to be supported.
+type Constraints struct {
+	Version *semver.Constraints
+}
+
+// Validate a Rule. Special attention is paid to fields of type
+// *semver.Constraints, because when parsing a Rule from JSON, a field of
+// this type that is not present is set to nil.
+func (rule Rule) Validate() error {
+	// validate Criteria.AppVersion
+	if rule.Criteria.AppVersion == nil {
+		return fmt.Errorf("invalid Criteria.AppVersion %q", rule.Criteria.AppVersion)
+	}
+
+	// validate Criteria.Platform
+	if !validPlatform[rule.Criteria.Platform] {
```

Comment:
I revisited this some time ago and came to the conclusion that it would be better to:

1. Ignore `Criteria.PlatformVersion` for Linux, since the concept doesn't apply to Linux (because it's a kernel).
2. Require `Criteria.Platform` to be specified. My (faulty) thinking was that we will always want to specify `Criteria.Platform` in a rule. The thing that influenced my thinking here was that it would be possible to specify `Criteria.PlatformVersion` without specifying any constraint for `Criteria.Platform`. This means we could require a certain OS version, but not know which OS that OS version was for. Obviously this wouldn't make sense.

I've reverted this for the following reasons:

For point 1, we are now sending `0.0.0` for `PlatformVersion` from the client, so we don't have to worry about parsing issues (previously we were sending `"unknown"`). Also, I think it is reasonable to trust the Upgrade Responder admins to know enough to not specify `PlatformVersion` for Linux `Rule`s.

Point 2 is solved by checking another case in validation: if `Criteria.PlatformVersion` is not `*`, then `Criteria.Platform` cannot be `*`. This prevents the ambiguity described above while letting us specify rules that apply to all platforms.

Hopefully that sounds alright to you! Let me know if you want me to change anything.

Reply:
This all sounds good. I still would like to see the spec imported into a Markdown document inside the repo, but that can be a separate PR.